### PR TITLE
Add Streaming Dataset

### PR DESCRIPTION
**Pull request by fracapuano**
More details around streaming in #1173

## What this does

This PR introduces two key components:
1. A new `StreamingLeRobotDataset` class for efficient streaming of robot datasets
2. A profiling script to measure and analyze its performance characteristics

## `StreamingLeRobotDataset`

The `StreamingLeRobotDataset` class extends the standard dataset functionality with streaming capabilities, allowing data to be processed without loading the entire dataset into memory.

### Key Features:
- **Memory-efficient streaming**: Loads data on-demand rather than all at once
- **Buffer-based sampling**: Maintains a bounded buffer of recent items for efficient access
- **Sharded loading**: Supports loading data from multiple shards for improved performance and randomness
- **Video frame decoding**: Efficiently handles video data downloading only certain frames in a video. Uses caching for improved performance (~10x)
- **Random sampling**: Implements efficient random sampling from the stream

## Profiling Script

The accompanying profiling script (`profile_streaming_dataset.py`) provides detailed performance analysis. **Note: `profile_streaming_dataset.py` is not intended for merge and will be removed when this PR is ready to merge)**

### Features:
- **Detailed timing measurements**: Per-sample and per-run timing statistics
- **Warmup support**: Configurable warmup iterations to measure overhead
- **Line profiling**: Detailed line-by-line profiling of key dataset methods
- **Progress visualization**: Uses tqdm to show progress during profiling runs
- **Comprehensive reporting**: Saves results to a file for further analysis

### Streaming Dataset:
```python
from lerobot.common.datasets.streaming_dataset import StreamingLeRobotDataset

dataset = StreamingLeRobotDataset(
    repo_id="lerobot/aloha_mobile_cabinet",
    buffer_size=1000,
    max_num_shards=16
)

for item in dataset:
    # Process item
    pass
```

### Profiling:
```bash
python -m lerobot.common.datasets.profile_streaming_dataset \
    --repo-id lerobot/aloha_mobile_cabinet
```

## How to checkout & try? (for the reviewer)
Checkout to this branch and use the aforementioned snippets. Unsurprisingly, one needs a tiny bit of version management to make sure everything works 😅 In particular, the latest version of `torchcodec` (enabling single-frame streaming) requires specific dependancies, namely you might need to reinstall/update `torch`, `torchvision`, `torchaudio` and `ffmpeg`. Here are the versions I am using to run this locally:

```bash
torch                                2.7.0
torchaudio                           2.7.0
torchcodec                           0.4.0
torchvision                          0.22.0
imageio                              2.37.0  # ffmpeg
```

### CODE REVIEW

File: src/lerobot/datasets/streaming_dataset.py [hunk 1]
```diff
@@ -0,0 +1,543 @@
...
+        # loads 4 images: 1 second before current frame, 500 ms before, 200 ms before, and current frame
+        camera_key: [-1, -0.5, -0.20, 0],
+        # loads 6 state vectors: 1.5 seconds before, 1 second before, ... 200 ms, 100 ms, and current frame
+        "observation.state": [-1.5, -1, -0.5, -0.20, -0.10, 0],
+        # loads 64 action vectors: current frame, 1 frame in the future, 2 frames, ... 63 frames in the future
+        "action": [t / fps for t in range(64)],
+    }
+
+    dataset = StreamingLeRobotDataset(repo_id, delta_timestamps=delta_timestamps)
+    frame = next(iter(dataset))
```
Comment (michel-aractingi) [inline src/lerobot/datasets/streaming_dataset.py]: Remove the main from this script or add as an example script

File: src/lerobot/datasets/utils.py [hunk 2]
```diff
@@ -770,3 +773,224 @@ def to_parquet_with_hf_images(df: pandas.DataFrame, path: Path) -> None:
...
+
+class LookAheadError(Exception):
+    """
+    Exception raised when trying to look ahead in the future of a Backtrackable object.
+    """
+
+    pass
+
+
+class Backtrackable(Generic[T]):
```
Comment (michel-aractingi) [inline src/lerobot/datasets/utils.py:829]: Seems these can fit in a seperate `streaming_utils.py` script

File: src/lerobot/scripts/train.py [hunk 3]
```diff
@@ -208,6 +209,9 @@ def train(cfg: TrainPipelineConfig):
 
         for key in batch:
             if isinstance(batch[key], torch.Tensor):
+                if batch[key].dtype != torch.bool:
+                    batch[key] = batch[key].type(torch.float32) if device.type == "mps" else batch[key]
+
```
Comment (michel-aractingi) [inline src/lerobot/scripts/train.py:214]: What are these lines for?

File: src/lerobot/scripts/train.py [hunk 4]
```diff
@@ -179,10 +179,11 @@ def train(cfg: TrainPipelineConfig):
         dataset,
         num_workers=cfg.num_workers,
         batch_size=cfg.batch_size,
-        shuffle=shuffle,
+        shuffle=shuffle and not cfg.dataset.streaming,
         sampler=sampler,
         pin_memory=device.type == "cuda",
         drop_last=False,
+        prefetch_factor=2,
```
Comment (michel-aractingi) [inline src/lerobot/scripts/train.py:186]: why did you modify the prefetch factor here?

File: src/lerobot/scripts/train.py [hunk 5]
```diff
@@ -179,10 +179,11 @@ def train(cfg: TrainPipelineConfig):
         dataset,
         num_workers=cfg.num_workers,
         batch_size=cfg.batch_size,
-        shuffle=shuffle,
+        shuffle=shuffle and not cfg.dataset.streaming,
```
Comment (michel-aractingi) [inline src/lerobot/scripts/train.py:182]: Is this the only addition to get train.py to work with streaming?

File: src/lerobot/datasets/streaming_dataset.py [hunk 6]
```diff
@@ -0,0 +1,543 @@
...
+
+        self.streaming = streaming
+        self.buffer_size = buffer_size
+
+        # We cache the video decoders to avoid re-initializing them at each frame (avoiding a ~10x slowdown)
+        self.video_decoder_cache = None
+
+        # Unused attributes
+        self.image_writer = None
+        self.episode_buffer = None
```
Comment (michel-aractingi) [inline src/lerobot/datasets/streaming_dataset.py]: Defined but not used anywhere,

File: src/lerobot/datasets/streaming_dataset.py [hunk 7]
```diff
@@ -0,0 +1,543 @@
...
+                StopIteration,
+            ):  # NOTE: StopIteration inside a generator throws a RuntimeError since python 3.7
+                del idx_to_backtrack_dataset[shard_key]  # Remove exhausted shard, onto another shard
+
+        # Once shards are all exhausted, shuffle the buffer and yield the remaining frames
+        rng.shuffle(frames_buffer)
+        yield from frames_buffer
+
+    def _get_window_steps(
+        self, delta_timestamps: dict[list[float]] | None = None, dynamic_bounds: bool = False
```
Comment (michel-aractingi) [inline src/lerobot/datasets/streaming_dataset.py]: type nit maybe?
```suggestion
        self, delta_timestamps: dict[str, list[float]] | None = None, dynamic_bounds: bool = False
```

File: src/lerobot/datasets/streaming_dataset.py [hunk 8]
```diff
@@ -0,0 +1,543 @@
...
+    ) -> tuple[int, int]:
+        if delta_timestamps is None:
+            return 1, 1
+
+        if not dynamic_bounds:
+            # Fix the windows
+            lookback = 100
+            lookahead = 100
```
Comment (michel-aractingi) [inline src/lerobot/datasets/streaming_dataset.py]: should we put these constants as default global variables?

File: src/lerobot/datasets/streaming_dataset.py [hunk 9]
```diff
@@ -0,0 +1,543 @@
...
+
+            try:
+                for frame in self.make_frame(backtrack_dataset):
+                    if len(frames_buffer) == self.buffer_size:
+                        i = next(buffer_indices_generator)  # samples a element from the buffer
+                        yield frames_buffer[i]
+                        frames_buffer[i] = frame
+                    else:
+                        frames_buffer.append(frame)
+                    break  # random shard sampled, switch shard
```
Comment (michel-aractingi) [inline src/lerobot/datasets/streaming_dataset.py:221]: I am abit confused as to the job of this `break`. does it mean that from each shard we are only taking one sample?

Maybe more comments are needed here

File: src/lerobot/datasets/streaming_dataset.py [hunk 10]
```diff
@@ -0,0 +1,543 @@
...
+        episode_boundaries_ts: dict[str, tuple[float, float]] | None = None,
+    ) -> dict[str, list[float]]:
+        query_timestamps = {}
+        keys_to_timestamps = self._make_timestamps_from_indices(current_ts, query_indices)
+        for key in self.meta.video_keys:
+            if query_indices is not None and key in query_indices:
+                timestamps = keys_to_timestamps[key]
+                # Clamp out timesteps outside of episode boundaries
+                query_timestamps[key] = torch.clamp(
+                    torch.tensor(timestamps), *episode_boundaries_ts[key]
```
Comment (michel-aractingi) [inline src/lerobot/datasets/streaming_dataset.py:375]: are you sure we need `torch.tensor` here? Seems `_make_timestamps_from_indices` already has this casting.

File: src/lerobot/datasets/streaming_dataset.py [hunk 11]
```diff
@@ -0,0 +1,543 @@
...
+            video_path = f"{root}/{self.meta.get_video_file_path(ep_idx, video_key)}"
+            frames = decode_video_frames_torchcodec(
+                video_path, query_ts, self.tolerance_s, decoder_cache=self.video_decoder_cache
+            )
+
+            item[video_key] = frames.squeeze(0) if len(query_ts) == 1 else frames
+
+        return item
+
+    def _make_padding_frame(self, key: str) -> tuple[torch.Tensor, bool]:
```
Comment (michel-aractingi) [inline src/lerobot/datasets/streaming_dataset.py]: Unused function, remove?

File: src/lerobot/datasets/streaming_dataset.py [hunk 12]
```diff
@@ -0,0 +1,543 @@
...
+                    start_ts + torch.tensor(indices[key]) / self.fps
+                ).tolist()  # NOTE: why not delta_timestamps directly?
+                for key in self.delta_timestamps
+            }
+        else:
+            return dict.fromkeys(self.meta.video_keys, [start_ts])
+
+    def _make_padding_camera_frame(self, camera_key: str):
+        """Variable-shape padding frame for given camera keys, given in (C, H, W)"""
+        return torch.zeros(self.meta.info["features"][camera_key]["shape"]).permute(-1, 0, 1)
```
Comment (michel-aractingi) [inline src/lerobot/datasets/streaming_dataset.py:272]: why do we want to permute the camera frames from CHW -> W,C,H

File: src/lerobot/datasets/streaming_dataset.py [hunk 13]
```diff
@@ -0,0 +1,543 @@
...
+
+    def _get_video_frame_padding_mask(
+        self,
+        video_frames: dict[str, torch.Tensor],
+        query_timestamps: dict[str, list[float]],
+        original_timestamps: dict[str, list[float]],
+    ) -> tuple[dict[str, torch.BoolTensor]]:
```
Comment (michel-aractingi) [inline src/lerobot/datasets/streaming_dataset.py]: Typing : Seems that you return a dict not a tuple?

File: src/lerobot/datasets/utils.py [hunk 14]
```diff
@@ -770,3 +773,224 @@ def to_parquet_with_hf_images(df: pandas.DataFrame, path: Path) -> None:
...
+    Returns:
+        dict: Dictionary with all tensor-like items converted to torch.Tensor.
+    """
+    import numpy as np
+    import torch
+
+    for key, val in item.items():
+        if isinstance(val, (np.ndarray, list)) and key not in ["task"]:
+            # Convert numpy arrays and lists to torch tensors
+            item[key] = torch.tensor(val)
```
Comment (michel-aractingi) [inline src/lerobot/datasets/utils.py:798]: if we only have numpy arrays calling from_numpy would be better to avoid copying the data

File: src/lerobot/datasets/utils.py [hunk 15]
```diff
@@ -770,3 +773,224 @@ def to_parquet_with_hf_images(df: pandas.DataFrame, path: Path) -> None:
...
+    This function is used to convert an item from a streaming dataset to PyTorch tensors.
+
+    Args:
+        item (dict): Dictionary of items from a dataset.
+
+    Returns:
+        dict: Dictionary with all tensor-like items converted to torch.Tensor.
+    """
+    import numpy as np
+    import torch
```
Comment (michel-aractingi) [inline src/lerobot/datasets/utils.py]: We don't need to have these imports here, move to module level

File: src/lerobot/datasets/video_utils.py [hunk 16]
```diff
@@ -172,15 +174,66 @@ def decode_video_frames_torchvision(
...
+        if importlib.util.find_spec("torchcodec"):
+            from torchcodec.decoders import VideoDecoder
+        else:
+            raise ImportError("torchcodec is required but not available.")
+
+        video_path = str(video_path)
+
+        with self._lock:
+            if video_path not in self._cache:
+                file_handle = fsspec.open(video_path).__enter__()
```
Comment (michel-aractingi) [inline src/lerobot/datasets/video_utils.py:192]: since we are calling open and enter here, shouldn't we call exit or close after to be safe?

**Review by michel-aractingi — COMMENTED** [review]
Left a small review, great work @fracapuano !

File: src/lerobot/datasets/streaming_dataset.py [hunk 17]
```diff
@@ -0,0 +1,543 @@
+from collections.abc import Callable, Generator, Iterator
```
Comment (michel-aractingi) [inline src/lerobot/datasets/streaming_dataset.py:16]: Missing lisence

File: tests/datasets/test_streaming.py [hunk 18]
```diff
@@ -0,0 +1,375 @@
+import numpy as np
```
Comment (michel-aractingi) [inline tests/datasets/test_streaming.py:16]: Missing Lisence

File: src/lerobot/datasets/streaming_dataset.py (hunk 17 above)
Comment (fracapuano) [reply src/lerobot/datasets/streaming_dataset.py:16]: added

File: src/lerobot/datasets/streaming_dataset.py (hunk 6 above)
Comment (fracapuano) [reply src/lerobot/datasets/streaming_dataset.py]: removed

File: src/lerobot/datasets/streaming_dataset.py (hunk 9 above)
Comment (fracapuano) [reply src/lerobot/datasets/streaming_dataset.py:221]: yes. the logic is to add 2 levels of randomness: (1) sample one shard at random from the ones available and (2) sample one frame from the shard sampled

Will add this comment

File: src/lerobot/datasets/streaming_dataset.py (hunk 7 above)
Comment (fracapuano) [reply src/lerobot/datasets/streaming_dataset.py]: ty. fixed

File: src/lerobot/datasets/streaming_dataset.py (hunk 8 above)
Comment (fracapuano) [reply src/lerobot/datasets/streaming_dataset.py]: added

File: src/lerobot/datasets/streaming_dataset.py (hunk 13 above)
Comment (fracapuano) [reply src/lerobot/datasets/streaming_dataset.py]: fixed, thank you

File: src/lerobot/datasets/streaming_dataset.py (hunk 11 above)
Comment (fracapuano) [reply src/lerobot/datasets/streaming_dataset.py]: removed, ty

File: src/lerobot/datasets/streaming_dataset.py (hunk 1 above)
Comment (fracapuano) [reply src/lerobot/datasets/streaming_dataset.py]: done, thank you

File: src/lerobot/datasets/utils.py (hunk 15 above)
Comment (fracapuano) [reply src/lerobot/datasets/utils.py]: done

File: src/lerobot/datasets/utils.py (hunk 2 above)
Comment (fracapuano) [reply src/lerobot/datasets/utils.py:829]: I'd rather keep it in here. I see `StreamingLeRobotDataset` and `LeRobotDataset` becoming one Dataset only (like datasets) and I think this helps going there (plus, is clearer to have as we don't need to add that many files)

File: src/lerobot/scripts/train.py (hunk 5 above)
Comment (fracapuano) [reply src/lerobot/scripts/train.py:182]: yes, :)

File: src/lerobot/scripts/train.py (hunk 4 above)
Comment (fracapuano) [reply src/lerobot/scripts/train.py:186]: I didn't, I added it. I don't think it was there before

File: src/lerobot/scripts/train.py (hunk 3 above)
Comment (fracapuano) [reply src/lerobot/scripts/train.py:214]: boolean tensors on mps are cumbersome to use this makes everything work

File: src/lerobot/datasets/video_utils.py (hunk 16 above)
Comment (fracapuano) [reply src/lerobot/datasets/video_utils.py:192]: added that to the cache clearing process. Problem with doing that always is that we want it to be open until then because we're caching the _connector_ to the video rather than the video itself (hence the connection'd have to be open)

File: src/lerobot/datasets/streaming_dataset.py [hunk 19]
```diff
@@ -0,0 +1,535 @@
...
+                key: (
+                    start_ts + torch.tensor(indices[key]) / self.fps
+                ).tolist()  # NOTE: why not delta_timestamps directly?
+                for key in self.delta_timestamps
+            }
+        else:
+            return dict.fromkeys(self.meta.video_keys, [start_ts])
+
+    def _make_padding_camera_frame(self, camera_key: str):
+        """Variable-shape padding frame for given camera keys, given in (C, H, W)"""
```
Comment (michel-aractingi) [inline src/lerobot/datasets/streaming_dataset.py]:
```suggestion
        """Variable-shape padding frame for given camera keys, given in (H, W, C)"""
```

File: src/lerobot/datasets/streaming_dataset.py (hunk 10 above)
Comment (michel-aractingi) [reply src/lerobot/datasets/streaming_dataset.py:375]: According to Francesco this is generally floats
